src/arch/riscv/pmp.hh

```diff
@@ -0,0 +1,194 @@
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __ARCH_RISCV_PMP_HH__
+#define __ARCH_RISCV_PMP_HH__
+
+#include "arch/generic/tlb.hh"
+#include "arch/riscv/isa.hh"
+#include "base/addr_range.hh"
+#include "base/types.hh"
+#include "mem/packet.hh"
+#include "params/PMP.hh"
+#include "sim/sim_object.hh"
+
+/**
+ * @file
+ * PMP header file.
+ */
+
+/**
+ * This class helps to implement RISCV's physical memory
+ * protection (pmp) primitive.
+ * @todo Add statistics and debug prints.
+ */
+class PMP : public SimObject
+{
+  public:
+    typedef PMPParams Params;
+
+    const Params &
+    params() const
+    {
+        return dynamic_cast<const Params &>(_params);
+    }
+
+    PMP(const Params &params);
+
+  private:
+    /** maximum number of entries in the pmp table */
+    int maxEntries;
+
+    /** This enum is used for encoding of address matching mode of
+     * pmp address register, which is present in bits 3-4 (A) of
+     * pmpcfg register for a pmp entry.
+     * PMP_OFF = null region (pmp disabled)
+     * MP_TOR = top of range mode
+     * PMP_NA4 = naturally aligned four byte region
+     * PMP_NAPOT = naturally aligned power of two region, >= 8 bytes
+     */
+    typedef enum {
+        PMP_OFF,
+        PMP_TOR,
+        PMP_NA4,
+        PMP_NAPOT
+    } pmpAmatch;
+
+    /** pmpcfg address range read permission mask */
+    const uint8_t PMP_READ = 1 << 0;
+
+    /** pmpcfg address range write permission mask */
+    const uint8_t PMP_WRITE = 1 << 1;
+
+    /** pmpcfg address range execute permission mask */
+    const uint8_t PMP_EXEC = 1 << 2;
+
+    /** pmpcfg address range locked mask */
+    const uint8_t PMP_LOCK = 1 << 7;
+
+    /** variable to keep track of active number of rules any time */
+    int numRules;
+
+    /** single pmp entry struct*/
+    typedef struct {
+        /** addr range corresponding to a single pmp entry */
+        AddrRange pmpAddr = AddrRange(-1, -2);
```

Comment:
I suggest initializing this to `AddrRange(0, 0)` I'm not sure what a negative address means. I think it's a uint64_t, so these are actually just big numbers. (Or, they could be ints cast to uint64_t in which case it's 0xFFFFFFFF).

(0,0) is an empty range, so it won't match anything

Reply:
fixed to (0,0). There was no need to have negative addresses. I think I was just trying to debug something and though that I am having unintended matches for address 0, so changed the defaults to negative addresses and did not revert them later.